File: demo/src/pingit/score.ts
// Faithful TypeScript port of the SDK's score/Score.kt — identical weights,
// normalization, and label buckets, so the demo scores exactly like the SDK.
import type { Metrics } from './types';

/** "more is better" → 0..1, reaching 1.0 at `full`. */
function saturating(value: number, full: number): number {
  if (value <= 0) return 0;
  return Math.min(1, value / full);
}

/** "less is better" → 1.0 at/below `good`, 0.0 at/above `bad`, linear between. */
function decreasing(value: number, good: number, bad: number): number {
  if (value <= good) return 1;
  if (value >= bad) return 0;
  return Math.max(0, 1 - (value - good) / (bad - good));
}

/** Weighted 0..100 score (download .35, upload .15, latency .25, jitter .10, loss .15). */
export function score(m: Metrics): number {
  const combined =
    0.35 * saturating(m.downloadMbps, 25) +
    0.15 * saturating(m.uploadMbps, 10) +
    0.25 * decreasing(m.latencyMs, 20, 300) +
    0.1 * decreasing(m.jitterMs, 5, 60) +
    0.15 * decreasing(m.packetLossPct, 0, 10);
  // Kotlin `.toInt()` truncates toward zero.
  return Math.min(100, Math.max(0, Math.trunc(combined * 100)));
}

/** Bucket a score into a human label (matches the SDK). */
export function label(s: number): string {
  if (s < 40) return 'Poor';
  if (s < 60) return 'Fair';
  if (s < 85) return 'Good';
  return 'Excellent';
}

File: sdk-android/core/src/main/kotlin/com/pingit/core/score/Score.kt
package com.pingit.core.score

import kotlin.math.max
import kotlin.math.min

object Score {

    /** Raw measured metrics used to compute a quality score. */
    data class Metrics(
        val download: Double,
        val upload: Double,
        val latency: Double,
        val jitter: Double,
        val loss: Double,
    )

    /**
     * Weighted 0..100 score. Each sub-metric is normalized to a 0..1 sub-score,
     * combined with weights, then scaled and clamped.
     *
     * Weights: download 0.35, upload 0.15, latency 0.25, jitter 0.10, loss 0.15.
     */
    fun score(m: Metrics): Int {
        val downloadScore = saturating(m.download, full = 25.0)
        val uploadScore = saturating(m.upload, full = 10.0)
        val latencyScore = decreasing(m.latency, good = 20.0, bad = 300.0)
        val jitterScore = decreasing(m.jitter, good = 5.0, bad = 60.0)
        val lossScore = decreasing(m.loss, good = 0.0, bad = 10.0)

        val combined =
            0.35 * downloadScore +
                0.15 * uploadScore +
                0.25 * latencyScore +
                0.10 * jitterScore +
                0.15 * lossScore

        return (combined * 100.0).toInt().coerceIn(0, 100)
    }

    /** Map a "more is better" metric to 0..1, reaching 1.0 at [full]. */
    private fun saturating(value: Double, full: Double): Double {
        if (value <= 0.0) return 0.0
        return min(1.0, value / full)
    }

    /**
     * Map a "less is better" metric to 0..1: 1.0 at or below [good], 0.0 at or
     * above [bad], linear in between.
     */
    private fun decreasing(value: Double, good: Double, bad: Double): Double {
        if (value <= good) return 1.0
        if (value >= bad) return 0.0
        return max(0.0, 1.0 - (value - good) / (bad - good))
    }

    /** Bucket a numeric score into a human label. */
    fun label(score: Int): String = when {
        score < 40 -> "Poor"
        score < 60 -> "Fair"
        score < 85 -> "Good"
        else -> "Excellent"
    }
}
